doc/poor_mans_mmp_design.txt

```diff
@@ -0,0 +1,70 @@
+MMP (Multi-Modifier Protection) attempts to prevent a user from
+importing a pool on more than one node at a time.
+
+See
+	https://github.com/zfsonlinux/zfs/issues/745
+and
+	https://github.com/zfsonlinux/zfs/pull/5821
+for more background.
+
+This design is for a very simple subset of MMP that can be coded,
+tested, and reviewed quickly.
+
+It attempts to detect and prevent the case of importing a pool via "zpool
+import -f", while it is already imported on another node.
+
+On-disk changes:
+
+The last uberblock slot in each ring is used only for MMP; uberblocks
+written by the sync thread always go into the first (N-1) slots.  The
+slot used for MMP is used to hold an uberblock which is exactly the
+same as the last written uberblock, except that its ub_timestamp is
+frequently updated, and its embedded checksum is different
+accordingly.
+
+This is used to provide the "heartbeat" without overwriting good
+uberblocks.
+
+During Import:
+
+zpool_do_import() (user space)
+1. Examine the hostid.  If it is 0,
+   skip the steps below and perform the import as normal.
+2. Issue the ZFS_IOC_POOL_TRYIMPORT ioctl
+3. Extract the txg and timestamp from the returned config nvlist and
+   save them.
+4. Sleep 1 second.
```

Comment:
For the archives: there is a use case where we build high-availability clusters. We also use fencing techniques outside of ZFS, such as SCSI reservations. In this use case, we never want to wait for a condition that cannot happen in the system because it needlessly increases the failover time budget. Hence, the desire to keep the feature tunable to the systems designers.

Similarly, there are some device types which, by design are not shared: eg SATA. So if someone is only using SATA, there is no need to wait.